File: package/android/src/main/java/com/margelo/nitro/realfetch/HybridInputStream.kt
package com.margelo.nitro.realfetch

import com.margelo.nitro.core.ArrayBuffer
import com.margelo.nitro.core.Promise
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import java.io.InputStream

class HybridInputStream(
  val stream: InputStream,
  private val onClose: (() -> Unit)? = null
) : HybridInputStreamSpec() {
  companion object {
    private const val BUFFER_SIZE = 64 * 1024
  }

  override fun read(): Promise<ArrayBuffer> {
    return Promise.async(CoroutineScope(Dispatchers.IO)) {
      val bytes = ByteArray(BUFFER_SIZE)
      val bytesRead = stream.read(bytes, 0, bytes.size)

      when {
        bytesRead == -1 -> {
          val emptyBuffer = ArrayBuffer.allocate(0)
          emptyBuffer
        }
        bytesRead > 0 -> {
          val arrayBuffer = ArrayBuffer.allocate(bytesRead)
          val destBuffer = arrayBuffer.getBuffer(false)
          destBuffer.put(bytes, 0, bytesRead)
          arrayBuffer
        }
        else -> {
            throw Error("Unexpected error reading stream")
        }
      }
    }
  }

  override fun open() {
  }

  override fun close() {
    try {
      stream.close()
    }
    finally {
      onClose?.invoke()
    }
  }

  override val memorySize: Long
    get() = try {
      stream.available().toLong()
    } catch (e: Exception) {
      0L
    }
}

File: package/ios/HybridInputStream.swift
import Foundation
import NitroModules

class HybridInputStream: HybridInputStreamSpec {
  let stream: InputStream

  init(stream: InputStream) {
    self.stream = stream
  }

  func open() throws {
    stream.open()
  }

  func read() throws -> Promise<ArrayBuffer> {
    let promise = Promise<ArrayBuffer>()

    Task {
      let size = Int(64 * 1024)
      let data = UnsafeMutablePointer<UInt8>.allocate(capacity: size)

      let bytesRead = stream.read(data, maxLength: size)

      let deleteFunc = {
        data.deallocate()
      }

      if bytesRead >= 0 {
        promise.resolve(
          withResult: ArrayBuffer.wrap(
            dataWithoutCopy: data, size: bytesRead, onDelete: deleteFunc))
      } else {
        deleteFunc()
        promise.reject(
          withError: stream.streamError
            ?? RuntimeError.error(withMessage: "Unexpected error reading stream"))
      }
    }

    return promise
  }

  func close() {
    stream.close()
  }
}
